### DISCUSSION TITLE
Matrix of ToggleButtons slow to load

### DISCUSSION THREAD

**Discussion by Ryan-K-94** [AvaloniaUI/Avalonia#12801] [Q&A] [answered]
One of the views in an application I'm working on is a matrix of toggle buttons nested inside a items repeater and scroll viewer. It's size is dynamic but typically around 64 x 64. The matrix is displayed in a tab item. When the tab is selected its taking quite a while to load the view (around 5 seconds). 

I've tried changing the toggle button to a border and this made the view load a lot quicker (< 1 second). Although the border is lacking a lot of features that I need so it's not the solution.

The matrix is part of a large project that I can't post but I've included the applicable XMAL and the data type it's bound to. Obviously some of the bindings are application specific.

I'm new to Avalonia, would greatly appreciate any suggestions that might make the view quicker to load. There's no problem with it's responsiveness when it's actually loaded. Is there a lighter weight alternative to the toggle button that will allow me to achieve the same thing. 

Many Thanks!

    public List<List<MatrixItem>>? Column { get; set; }

    public class MatrixItem : ReactiveObject
    {
        private bool isAssigned;
        private InOut isAssignedTooInOut;
        private MatrixButtonSate buttonState;

        public ICommand? Command { get; set; }
        public int Row { get; set; }
        public int Column { get; set; }
        public int SatBoxId { get; set; }

        public bool IsAssigned
        {
            get => isAssigned;
            set { this.RaiseAndSetIfChanged(ref isAssigned, value); }
        }

        public InOut IsAssignedTooInOut
        {
            get { return isAssignedTooInOut; }
            set { this.RaiseAndSetIfChanged(ref isAssignedTooInOut, value); }
        }

        public MatrixButtonSate ButtonState
        {
            get { return buttonState; }
            set { this.RaiseAndSetIfChanged(ref buttonState, value); }
        }
    }

	<ScrollViewer HorizontalScrollBarVisibility="Disabled" VerticalScrollBarVisibility="Visible" Margin="0,0,0,10">
			
			<ItemsRepeater Items="{Binding PatchMatrix.MatrixColumn, Mode=TwoWay}">
				<ItemsRepeater.Layout>
					<StackLayout Spacing="5" Orientation="Horizontal" />
				</ItemsRepeater.Layout>
				<ItemsRepeater.ItemTemplate>
					<DataTemplate>
						<SplitView Name="RootSplitView0" IsPaneOpen="{Binding IsPaneOpen}" DisplayMode="CompactInline" PaneBackground="Transparent"
								   OpenPaneLength="{Binding OpenPaneLength, Mode=TwoWay}" CompactPaneLength="29">
							<SplitView.Pane>
									
									<ItemsControl Grid.Row="2" Name="GridItem0" Items="{Binding Column, Mode=TwoWay}" ItemTemplate="{DynamicResource DataTemplate_Level1}"
												  IsVisible="{Binding IsPaneOpen, Mode=TwoWay}" Width="{Binding OpenPaneLength, Mode=OneWayToSource}"/>
							</SplitView.Pane>
						</SplitView>
					</DataTemplate>
				</ItemsRepeater.ItemTemplate>
			</ItemsRepeater>
		</Grid>
	</ScrollViewer>

	<DataTemplate x:Key="DataTemplate_Level2">
		<ToggleButton Name="MatrixButton" HorizontalContentAlignment="Center" VerticalContentAlignment="Center" FontSize="8" Padding="0" Height="27" Width="27" Margin="1,1,1,1" 
					  Command="{Binding Command}" IsChecked="{Binding IsAssigned, Mode=TwoWay}" Background="{Binding ButtonState, Converter={StaticResource matrixBtnIsActiveColourConverter}}" 
					  PointerEnter="MatrixPointerEnter" PointerLeave="MatrixPointerLeave" >
					  
			<ToggleButton.CommandParameter>
				<MultiBinding Converter="{StaticResource matrixButtonIndexConverter}">
					<Binding Path="Row"/>
					<Binding Path="Column"/>
					<Binding Path="SatBoxId"/>
				</MultiBinding>
			</ToggleButton.CommandParameter>

			<ToggleButton.Styles>
				<Style Selector="ToggleButton:checked /template/ ContentPresenter">
					<Setter Property="Background" Value="{Binding IsAssignedTooInOut, Converter={StaticResource matrixToggleBtnColourConverter}}"/>
					<Setter Property="Opacity" Value="0.8"/>
				</Style>
			</ToggleButton.Styles>
			
		</ToggleButton>
	</DataTemplate>

	<DataTemplate x:Key="DataTemplate_Level1">
		<ItemsControl Items="{Binding}" ItemTemplate="{DynamicResource DataTemplate_Level2}">
			<ItemsControl.ItemsPanel>
				<ItemsPanelTemplate>
					<StackPanel Orientation="Horizontal"/>
				</ItemsPanelTemplate>
			</ItemsControl.ItemsPanel>
		</ItemsControl>
	</DataTemplate>

**Comment by timunie** [comment, marked as answer]
looks like you're using 0.10. Can you try your sample using 11.0?

**Reply by Ryan-K-94** [reply]
Thanks for the response. 

I had read that V11.0 comes with some performance improvements. I started taking a look at upgrading the project yesterday although it doesn't look like a small task. 

I will create a V11.0 application with just the matrix to test it if it's likely to solve the problem.

**Reply by Ryan-K-94** [reply]
Stand alone V11.0 test application is loading the matrix a lot quicker (around a second vs the 5 or so in 0.10). Have begun upgrading the main project.

**Reply by timunie** [reply]
In 11.0 you can specify your own `ControlTheme` swapping out the original style completely. This gives you the opportunity to provide a light-weight style that has probably less overhead and improves the performance on top of it. 

Glad to hear 11.0 has such a huge benefit for your :-)

**Reply by Ryan-K-94** [reply]
Great thanks! I will take a look.
